### ISSUE TITLE
RDM discovery not sending any serial messages, but OLA startup does

### ISSUE THREAD

**Issue by ssilverman** [OpenLightingProject/ola#1694]
It seems that RDM discovery sends no traffic to the USB widget. I have a serial monitor running. When OLA starts (`ola_dev_info` after no OLA is running (`killall olad`)), I see the following messages being sent from OLA to my serial-connected widget. Note that this is a custom device, and for testing purposes, I changed the device parameters to Manufacturer ID=0x6A6B (DMXking), serial=01020304, Device ID=4, and firmware=2.4, so that `WidgetDetectorThread` would create an RDM-capable `EnttecUsbProWidget` device.
```
DeviceManufacturer
DeviceName
GetSerial
GetParams
GetParams
SendRDM (broadcast) (MUTE)
SendRDM (broadcast) (MUTE)
SendRDM (broadcast) (MUTE)
SendRDMDiscovery
```

I have one RDM responder hooked up and I do get a response (Device UID=0x7FF001020304), which the widget then sends back to OLA:
```
packet was RDM Discovery response
Sent 30 bytes to host
packet[30]:
    7e 05 19 00
    00
    fe fe fe fe fe fe fe aa ff 7f fa f5 ab 55 aa 57 ab 57 ae 55 af 57 fb 77
    e7
```

But there's no further RDM messages, even when I do full or incremental RDM discovery. Unless I receive or send DMX (both of which work fine), I see no RDM serial traffic unless either `olad` is restarted or OLA periodically repeats the above messages.

What could explain there being no RDM traffic from OLA when requested? Am I misunderstanding how this works?

**Comment by peternewman** [comment]
Can you share olad -l 4 logs please:
https://www.openlighting.org/ola/get-help/ola-faq/#How_do_I_get_olad_-l_4_logs

> Note that this is a custom device, and for testing purposes, I changed the device parameters to Manufacturer ID=0x6A6B (DMXking), serial=01020304, Device ID=4, and firmware=2.4, so that `WidgetDetectorThread` would create an RDM-capable `EnttecUsbProWidget` device.

I think offhand you don't need to hijack the manufacturer, as long as you've got some labels and a sufficient firmware version you should get RDM capabilities I think.

> I have one RDM responder hooked up and I do get a response (Device UID=0x7FF001020304), which the widget then sends back to OLA:

Do you mean your widget gets a response?

> But there's no further RDM messages, even when I do full or incremental RDM discovery. Unless I receive or send DMX (both of which work fine), I see no RDM serial traffic unless either `olad` is restarted or OLA periodically repeats the above messages.
> 
> What could explain there being no RDM traffic from OLA when requested? Am I misunderstanding how this works?

It looks like you're probably not sending the RDM_TIMEOUT label. See e.g.:
https://github.com/vanvught/rpidmx512/blob/ac9beaada1d67c793760593e623d2299d7774585/lib-widget/src/widget.c#L425-L443

I thought @vanvught raised an issue but I can't find it.

Although if you don't want to implement this, there's another option which should be compatible with anything that can support the Enttec DMX USB Pro Mk 2 - #1631 .

**Comment by ssilverman** [comment]
Let me clarify a few things: the Prototype Manufacturer ID works too, 0x7FF0. I just chose the first obvious code path in `WidgetDetectorThread`. It's not a product and it's only local software on a custom device. I'm not hijacking anything. :)

Second, my setup looks like this:
OLA -> USB serial -> Enttec-compatible RDM widget device -> DMX connection -> RDM responder.

Glossary: When I say "widget", I'm referring to my custom Enttec-compatible controller hooked up over USB serial to OLA running on my Mac. When I say "responder", I mean the RDM responder device at the end of the chain, connected via DMX. I'll use "OLA" to mean "OLA software running on my Mac and talking over USB serial to my widget."

When I receive the DUB response from the responder (DMX packet starting with FE), I forward this over serial to OLA. However, OLA never sees this, or at least I don't think it does, because there's no results anywhere on the web interface nor RDM discovery places. I know the connection is working because DMX device (also running on the responder)->widget is successfully seen in the DMX Monitor page in the OLA web interface.

In other words, the widget sees an "RDM Discovery" message from OLA, it performs an RDM discovery, and then receives a correct packet from the responder, which I then forward back to OLA via a ReceivedDMXPacket (Label=5) message to OLA. (i.e. the DMX packet starting with 0xFE.)

When I run RDM discovery, either over the web interface or via the `ola_rdm_discover` command, I see absolutely no serial traffic generated by OLA on my widget.

**Comment by ssilverman** [comment]
I forgot to ask: Where is this RDM timeout label documented? It's nowhere in the Enttec protocol doc.
See: https://dol2kh495zr52.cloudfront.net/pdf/misc/dmx_usb_pro_api_spec.pdf

Are you saying that some undocumented "timeout" message needs to be sent from the widget to the host after an RDM Discovery (Label=11) is performed?

**Comment by ssilverman** [comment]
I'm experimenting with some things. Running `ola_rdm_discover` _never sends any serial traffic to the widget_, but I see:

When my widget returns firmware 0x040f (see https://github.com/OpenLightingProject/ola/blob/0028d658d6e58ec5158c017ae676f30103d75bbc/plugins/usbpro/WidgetDetectorThread.cpp#L326):
```
ola_rdm_discover -u 0 -l 4
common/io/SelectPoller.cpp:233: ss process time was 0.000000
```

When my widget returns firmware 0x0204: it hangs at the same command and keeps printing that `common/io/Select` statement.

**Comment by ssilverman** [comment]
Let me be even more specific with my questions:
1. If you're a widget and you receive a Discovery MUTE for broadcast from OLA (Label=7), how should the widget respond to OLA? I'm currently not responding to OLA if the MUTE message is broadcast, because broadcast messages shouldn't cause responses from responders.
2. If you're a widget and you receive an RDMDiscovery(Label=11) message from OLA, how should it respond to OLA once it gets a DUB response from the responder? I'm currently sending ReceivedDMXPacket(Label=5) after first sending out a Discovery request over DMX and receiving a response from a responder.

**Comment by peternewman** [comment]
> Let me clarify a few things: the Prototype Manufacturer ID works too, 0x7FF0. I just chose the first obvious code path in `WidgetDetectorThread`. It's not a product and it's only local software on a custom device. I'm not hijacking anything. :)

Cool, just thought it was worth checking. A few people have started out testing stuff with the OLP manufacturer ID and obviously when code gets out there it's very hard to undo it.

> When I receive the DUB response from the responder (DMX packet starting with 0xFE), I forward this over serial to OLA. However, OLA never sees this, or at least I don't think it does, because there's no results anywhere on the web interface nor RDM discovery places.

olad -l 4 will show the low level packets received (or at least the fact they existed if not always data) on the USB bus (Enttec protocol) even if the RDM discovery failed/RDM packet was corrupted.

> I know the connection is working because DMX device (also running on the responder)->widget is successfully seen in the DMX Monitor page in the OLA web interface.

Sorry, I don't follow, when you turn the setup around and use your responder as a DMX controller/transmitter/source?

> **This is my main question ->** When I run RDM discovery, either over the web interface or via the `ola_rdm_discover` command, I see absolutely no serial traffic generated by OLA on my widget. What is `ola_rdm_discover` supposed to do, if not communicate to the widget?

As per the man page:
http://docs.openlighting.org/ola/man/man1/ola_rdm_discover.1.html

> ola_rdm_discover - Fetch the UID list for a universe.

`ola_rdm_discover -u 1 -f`

> Force full RDM Discovery for this universe

I think most active discovery button presses in the web interface should generate a discovery command. Unless you've just switched to the RDM tab.

> **This is my second question ->** Tell me more about this Timeout (Label=12) message and its process/necessity? (Inclusive of why `no_rdm_dub_timeout` defaults to `false` everywhere.)

Prior to 4.14/4.15 they always sent an RDM timeout. At least some of the Mk 2 firmwares don't, hence we normally expect one from an Enttec aside from after those firmware versions, set here:
https://github.com/OpenLightingProject/ola/blob/0028d658d6e58ec5158c017ae676f30103d75bbc/plugins/usbpro/WidgetDetectorThread.cpp#L324-L328

**Comment by peternewman** [comment]
> I forgot to ask: Where is this RDM timeout label documented? It's nowhere in the Enttec protocol doc.
> See: https://dol2kh495zr52.cloudfront.net/pdf/misc/dmx_usb_pro_api_spec.pdf

Welcome to a half-documented closed source device...

See also:
https://github.com/OpenLightingProject/ola/blob/0028d658d6e58ec5158c017ae676f30103d75bbc/plugins/usbpro/EnttecUsbProWidget.cpp#L446-L448

Here's some proper history:
https://groups.google.com/g/open-lighting/c/EYphmbhTfTM/m/z7Ag95RTzZMJ

I can only guess their developers didn't document it internally either, which is why it dropped from the Mk 2 devices.

Consider for example a device which doesn't respond to a directed mute, as Simon says, without the widget sending a timeout we have to inaccurately time that within OLA.

> Are you saying that some undocumented "timeout" message needs to be sent from the widget to the host after an RDM Discovery (Label=11) is performed and after a "0xFE" response is received by the widget and passed to OLA?

Yes, essentially when there is no more RDM data.

> I'll add: Can you point me to docs that describe the procedure and required messages a widget needs for RDM Discovery, in order to interact with OLA? According to that spec doc, all a widget needs to do after receiving an RDM Discovery message from OLA is to respond with any received DMX packets (i.e. those starting with 0xFE).

Not really aside from the code sorry (or Arjun's widget code). It seems our Arduino example doesn't do discovery, it cheats.

> **Secondary question ->** Do the ENTTEC and DMXking RDM products themselves use this Timeout message?

Enttec's certainly do (prior to 4.14/4.15). I think DMXking does too, I'd need to test with mine or check the codepath, but I believe they do as well. I think Simon got Enttec to add the code, so I suspect he asked DMXking to mirror it too.

**Comment by peternewman** [comment]
> I'm experimenting with some things. Running `ola_rdm_discover` _never sends any serial traffic to the widget_.

As per the other message you need a `-f`.

> I found this:
> 
> https://github.com/OpenLightingProject/ola/blob/0028d658d6e58ec5158c017ae676f30103d75bbc/plugins/usbpro/WidgetDetectorThread.cpp#L326

Yeah as I linked to.

> When my widget returns firmware 0x040f, I see the same "no serial traffic and the command never ends" as I do when I use firmware 0x0204:

Again you need a `-f`.

> ```
> ola_rdm_discover -f -u 1 -l 4
> common/io/SelectPoller.cpp:233: ss process time was 0.000009
> common/io/SelectPoller.cpp:233: ss process time was 0.000006
> ....continues without end.
> ```
> 
> When I run `olad -l 4`, I see this when my widget receives the MUTE and RDMDiscover messages:

It would really help to have the full log, not just a snippet, to confirm if the device is /dev/ttyUSB0 and so is randomly dropping off. Also I suspect your 0x040f hasn't worked if you're still spoofing a DMXking widget, but the logs would confirm either way. Certainly OLA is still firing the timeout as it didn't get a timeout message either real or dummy from the widget.

Also which version of OLA are you running and from source or from git? The 0x040f thing is only in 0.10.8 which is still in Homebrew packaging (as per the link before).

> When I run an `ola_rdm_discovery -f -u 1 -l 4`, I only see this in the `olad` log:
> 
> `olad/plugin_api/Universe.cpp:523: Full RDM discovery triggered for universe 1`

I suspect it's sitting waiting for the timeout/or broken because it didn't get it.

> `common/io/Serial.cpp:151: Device /dev/ttyUSB0 doesn't exist, so there's no point trying to acquire a lock`

Or this is the spoof Enttec and it's gone away.

> There's nothing that says it's sending RDM packets.

Agreed.
 
> The DUB response message that the widget relays from the responder to OLA is (split up to make it easier to read):

Looks plausible.

> The only time I ever see `plugins/usbpro/EnttecUsbProWidget.cpp:582: Enttec Pro discovery complete:` in the `olad` log is when I press CTRL-C to stop the program.

I think it fires the callback on clean-up at programme exit.

**Comment by peternewman** [comment]
> Let me be even more specific with my questions:
> 
>     1. If you're a widget and you receive a Discovery MUTE for broadcast from OLA (Label=7), how should the widget respond to OLA? I'm currently not responding to OLA if the MUTE message is broadcast, because broadcast messages shouldn't cause responses from responders.

Off the top of my head I think if the widget is < 0x040f, it should send an RDM_TIMEOUT label to say nothing came on the DMX/RDM wire.

>     2. If you're a widget and you receive an RDMDiscovery(Label=11) message from OLA, how should it respond to OLA once it gets a DUB response from the responder? I'm currently sending ReceivedDMXPacket(Label=5) after first sending out a Discovery request over DMX and receiving a response from a responder.

Off the top of my head I think if the widget is < 0x040f, it should send an RDM_TIMEOUT label after its sent the actual data to say nothing more came on the DMX/RDM wire.

**Comment by ssilverman** [comment]
I got everything working. All your comments were very helpful, thank you.

Notes:
1. It’s always a good reminder to not use official manufacturer IDs. Even though I never saved the code nor committed it, future readers thank you for mentioning it. I’m glad OLA has a code path where are you don’t have to do that.
2. I was indeed not using 0.10.8, so the 0x040f firmware version didn’t do anything different. I’m using the 0.10.7_5 release in Homebrew.
3. I was adding the `-f` option when running the discovery command. OLA, it appears, doesn’t send any more messages unless it’s received an RDMTimeout. (Did this change in 0.10.8?)
4. May I suggest not waiting for timeout messages in OLA if the RDM message being sent is broadcast (non-discovery)? (Unless this is already considered or done in 0.10.8 or later.)
5. I’d also like to suggest that OLA implement a timeout if it doesn’t receive an expected RDMTimeout message from the widget. (How meta.) (Again, unless this has already been thought out.)

Thanks again for your time and your help.

**Comment by peternewman** [comment]
> I got everything working. All your comments were very helpful, thank you.

Glad to hear it. I don't suppose you could summarise where you needed to send the timeouts for other people's benefit, or are you planning on open sourcing the interface?

>     2. I was indeed not using 0.10.8, so the 0x040f firmware version didn’t do anything different. I’m using the 0.10.7_5 release in Homebrew.

Okay, well that makes sense then.

>     3. I was adding the `-f` option when running the discovery command. OLA, it appears, doesn’t send any more messages unless it’s received an RDMTimeout. (Did this change in 0.10.8?)

Ah apologies. No nothing in that regard has changed.

>     4. May I suggest not waiting for timeout messages in OLA if the RDM message being sent is broadcast (non-discovery)? (Unless this is already considered or done in 0.10.8 or later.)

It's also not changed in 0.10.8. The issue is even something as seemingly simple as this isn't as trivial as it first seems. For starters if you're actually doing this in the context of RDM testing, you want to know if a buggy responder incorrectly responds to a broadcast message. Even if you didn't, we need to stick to the behaviour of the numerous interfaces supported by this bit of the codebase. Without testing each one (on each firmware) there's no way to guarantee how they respond. It would be worse if they got hung up or dropped the next message because you sent them another message before they sent the timeout (as mentioned elsewhere in the code). I've only got two/three (and reasonably easy access to one or two more) of the thirteen devices this plugin supports and that's ignoring firmware versions. So I'm rather loath to take the risk of breaking stuff for a minor speed up in some circumstances.

>     5. I’d also like to suggest that OLA implement a timeout if it doesn’t receive an expected RDMTimeout message from the widget. (How meta.) (Again, unless this has already been thought out.)

It already does that. If you check the full log, you should find one of the following from where the watchdog fired:
https://github.com/OpenLightingProject/ola/blob/112f1f262812e0087c131ce4e393b35b0be35780/plugins/usbpro/EnttecUsbProWidget.cpp#L610-L633

It sounds like we perhaps don't reset correctly after this (I'd imagine plugging and unplugging the widget will make it start working again). Although from a quick look at the code it looks like it probably should do. If you can share a FULL olad -l 4 log with it firing one of those watchdog timeouts and then not sending anything on a subsequent discovery I can try and see what's going on.

Again see the notes just above for the complexities of "just" resetting better in this case.

Given as far as I'm aware this only happens with bugs in off the shelf responders, or DIY ones which don't follow the (unpublished) spec, the downsides would again seem to outweigh the minor benefits.

If someone can come up with a reproducible way to hang an Enttec and get it stuck in this mode then it would probably make sense to look at.

**Comment by ssilverman** [comment]
Hey, @peternewman. I haven’t forgotten about this and will get to answering your requests and questions. Thanks for your patience.

**Comment by ssilverman** [comment]
For the benefit of others, as you suggest, here's a summary of my timeout strategy:
1. If I'm in a state that is expecting an RDM or RDM Discovery packet and I haven't seen a packet within some timeout then I send that RDMTimeout message to the host (label=12).
2. If I'm in a state that's expecting an RDM Discovery packet and I receive a DUB response then I pass along that received packet and then send that RDMTimeout message (both to the host, consecutively).

> For starters if you're actually doing this in the context of RDM testing, you want to know if a buggy responder incorrectly responds to a broadcast message.

Makes sense and I understand that point of view.

> It sounds like we perhaps don't reset correctly after this

This might be the case because if my widget doesn't send an RDMTimeout then I see no more traffic. Here's a gist with the log: https://gist.github.com/ssilverman/eb09e28c58bc0ce47cd64469d127c7d5

Note that if my widget doesn't send any RDMTimeouts, and `olad` is started, I do see messages, but then never again. Commands like `ola_rdm_discover -u 1 -f` cause nothing to happen.
